core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -579,6 +582,8 @@ public PlanOptimizers(
                                 .addAll(columnPruningRules)
                                 .add(new InlineProjections(plannerContext, typeAnalyzer))
                                 .addAll(new PushFilterThroughCountAggregation(plannerContext).rules()) // must run after PredicatePushDown and after TransformFilteringSemiJoinToInnerJoin
+                                .addAll(new SimplifyFalseConditions(plannerContext).rules())
```

Comment:
There's already a rule for this: RemoveTrivialFilters. The new logic should be implemented within that rule.

Reply:
I think it was mainly because of @martint's comment from 4 years ago:
https://github.com/trinodb/trino/blob/2ddb5d8185c5c3724e5959f5675f8f2923686fe4/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPredicateIntoTableScan.java#L241-L243
But I don't object to merging these two, they seem to go together anyway + that will turn `RemoveTrivialFilters` into an `ExpressionRewriteRuleSet` so maybe it will also benefit the join case (I didn't check).
I'm not sure the term "trivial" will still be accurate though.

Reply:
It's ok to have a rule that simplifies expressions (we already have one) and one that removes filters where the expression is known to filter everything or nothing (we already have one, too). But we should not add yet another rule that mixes both concerns.

Reply:
I took another look, `RemoveTrivialFilters` can't become an `ExpressionRewriteRuleSet` because `ExpressionRewriteRuleSet` doesn't replace a node of one type with a node of another type. 
The new rule's logic can't be moved into `SimplifyExpressions` because not all node types are supported (see https://github.com/trinodb/trino/pull/16206#discussion_r1155329611).
I can make `RemoveTrivialFilters` return 2 rules, but I failed to prove the join case (couldn't find an actual query that has a final plan that contains a `JoinNode` with a filter that can be simplified to `FALSE` \ `TRUE`).
@findepi \ @martint WDYT should be the next step?

Reply:
> the join case (couldn't find an actual query that has a final plan that contains a JoinNode with a filter that can be simplified to FALSE \ TRUE).

I guess that's because PPD pushes such filter under the join and then it gets simplified/eliminated there

> But we should not add yet another rule that mixes both concerns.

I read it as implying `RemoveTrivialFilters` shouldn't call `DomainTranslator` for help.

We tried to introduce a generic rule `ExpressionRewriteRuleSet`-based to simplify boolean conditions that can be replaced with constant false. I know what problems this run into (distinguishing context: eg between projected). 
Did it result in more test coverage? like cases mentioned in https://github.com/trinodb/trino/pull/16206#discussion_r1168458969 ?

Reply:
> I read it as implying RemoveTrivialFilters shouldn't call DomainTranslator for help.

It's possible but I'm not sure, maybe we should wait for the commenter for clarification. Please note that adding this logic into `RemoveTrivialFilters` will address this concern of yours https://github.com/trinodb/trino/pull/16206#discussion_r1118616081.

> Did it result in more test coverage? like cases mentioned in https://github.com/trinodb/trino/pull/16206#discussion_r1168458969 ?

I didn't add new tests for those cases. The development process caused me to find out some issues that were fixed in separate PRs. There's also the other commit of this PR (1b34f0545fa22f5e17928d23a8451af52518aa62).
If we'll decide to undo this change, do you think the other commit is worth merging separately? I would also like to rephrase these comments 
https://github.com/trinodb/trino/blob/2ddb5d8185c5c3724e5959f5675f8f2923686fe4/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPredicateIntoTableScan.java#L241-L243

https://github.com/trinodb/trino/blob/e5732f8857f8e0a017fe015eb2fe12673a5e49b4/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantPredicateAboveTableScan.java#L110-L112
and explain why we don't want to do that logic in "another rule".

Reply:
> I read it as implying RemoveTrivialFilters shouldn't call DomainTranslator for help.

Please see https://github.com/trinodb/trino/pull/16206#discussion_r1169103398. Note that adding this logic into `RemoveTrivialFilters` will address this concern of yours https://github.com/trinodb/trino/pull/16206#discussion_r1118616081.

> Did it result in more test coverage? like cases mentioned in https://github.com/trinodb/trino/pull/16206#discussion_r1168458969 ?

I didn't add new tests for those cases.

Reply:
> I didn't add new tests for those cases.

Can you please do?
These are interesting test cases to have, regardless of the implementation choices we end up doing in this PR